netket/vqs/mc/mc_state/state.py

```diff
@@ -26,23 +26,37 @@
 
 from netket import jax as nkjax
 from netket import nn
+from netket.stats import Stats
+from netket.operator import AbstractOperator
 from netket.sampler import Sampler, SamplerState
 from netket.utils import maybe_wrap_module, deprecated, warn_deprecation, mpi, wrap_afun
 from netket.utils.types import PyTree, SeedT, NNInitFunc
 from netket.optimizer import LinearOperator
 from netket.optimizer.qgt import QGTAuto
 
-from netket.vqs import VariationalState
+from netket.vqs.base import VariationalState, expect, expect_and_grad
 
 
-def compute_chain_length(n_chains, n_samples):
+def compute_chain_length(n_chains, n_samples, chunk_size=None):
     if n_samples <= 0:
         raise ValueError("Invalid number of samples: n_samples={}".format(n_samples))
 
     chain_length = int(np.ceil(n_samples / n_chains))
+
+    if chunk_size is not None:
+        n_chains_per_rank = n_chains // mpi.n_nodes
+
+        n_chunks = int(np.ceil((chain_length * n_chains_per_rank) / chunk_size))
+        samples_per_rank = n_chunks * chunk_size
+        chain_length = int(samples_per_rank // n_chains_per_rank)
```

Comment:
We want to find a sample count s.t. both `n_chains` and `chunk_size` divide it, right? This logic doesn't do that, unless `n_chains_per_rank` divides `chunk_size` (or we get lucky), the last step yields a sample number that isn't a multiple of `chunk_size`. How about this?
```suggestion
    if chunk_size is None:
        chain_length = int(np.ceil(n_samples / n_chains))
    else:
        n_chains_per_rank = n_chains // mpi.n_nodes
        lot_size = np.lcm(n_chains_per_rank, chunk_size)
        n_lots = int(np.ceil(n_samples / lot_size))
        samples_per_rank = n_lots * lot_size
        chain_length = samples_per_rank // n_chains_per_rank
```
BTW, the `//` operator always returns an integer, no need to put it in `int()`

Reply:
Because we want to allow big chunks:
```suggestion
    samples_per_rank = n_samples // mpi.n_nodes
    if chunk_size is None or chunk_size > samples_per_rank:
        chain_length = int(np.ceil(n_samples / n_chains))
    else:
        n_chains_per_rank = n_chains // mpi.n_nodes
        lot_size = np.lcm(n_chains_per_rank, chunk_size)
        n_lots = int(np.ceil(n_samples / lot_size))
        samples_per_rank = n_lots * lot_size
        chain_length = samples_per_rank // n_chains_per_rank
```

Reply:
thanks for the review!
However, I should have deleted this code. This code path is not used at all. It's a leftover from older versions...

Reply:
Actually, some version of this would still be needed: if one sets 

```python
vs.n_samples = 1008
vs.chunk_size = 126
vs.n_samples = 1500
```
one should ensure that n_samples is a multiple of chunk size

Reply:
```suggestion
    chain_length = int(np.ceil(n_samples / n_chains))
    n_samples_new = chain_length * n_chains
    if n_samples_new != n_samples:
        warnings.warn(
            "n_samples={} does not divide n_chains={}, increased to {}".format(
                n_samples, n_chains, n_samples_new
            )
        )

    if chunk_size is not None:
        n_samples_per_rank = n_samples_new // mpi.n_nodes
        if chunk_size < n_samples_per_rank and n_samples_per_rank % chunk_size != 0:
            raise ValueError(
                f"chunk_size={chunk_size}`<`n_samples_per_rank={n_samples_per_rank}, "
                "chunk_size is not an integer fraction of `n_samples_per rank`. This is"
                "unsupported. Please change `chunk_size` so that it divides evenly the"
                "number of samples per rank."
            )
```